### ISSUE TITLE
"Catastrophic" JML internal error, Reason: Mismatched blocks

### ISSUE THREAD

**Issue by HansvdLaan** [OpenJML/OpenJML#559]
While doing some basic exercises for a course I'm following, I managed to get the following error:
_"A catastrophic JML internal error occurred.  Please report the bug with as much information as you can.
  Reason: MISMATCHED BLOCKS"_

The "full" error: 
![image](https://user-images.githubusercontent.com/9132846/30662816-b391a1a8-9e48-11e7-8e73-64815a59a514.png)

The code:

```
/* A simple linked list of entries. */
public class EntryList {
    /*@spec_public @*/ Object first;
    /*@spec_public @*/ EntryList rest;

    EntryList( Object first, EntryList rest) {
        this.first = first;
        this.rest = rest;
    }

    /*@ ensures (this.first == null && \result == 0) ||(this.rest == null && \result == 1)  ||
      @ (\result == rest.size() + 1);
      @*/
    public int size() {
        if(this.first == null) {
            return 0;
        }
        if(this.rest == null) {
            return 1;
        } else {
            return 1 + rest.size();
        }
    }
}

```

The stacktrace:
```
A catastrophic JML internal error occurred.  Please report the bug with as much information as you can.
  Reason: MISMATCHED BLOCKS
The system is out of resources.
Consult the following stack trace for details.
java.lang.StackOverflowError
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.visitSelect(JmlAssertionAdder.java:10409)
    at com.sun.tools.javac.tree.JCTree$JCFieldAccess.accept(JCTree.java:1897)
    at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    [...]
at com.sun.tools.javac.tree.TreeScanner.scan(TreeScanner.java:49)
    at org.jmlspecs.openjml.esc.JmlAssertionAdder.convertExpr(JmlAssertionAdder.java:1366)
ENDING with exit code 3


```

**Comment by davidcok** [comment]
Hans, Thanks for the bug report. The stack overflow should not happen of course, but here is a workaround until I get it fixed.

The root problem is the combination of (a) using a recursive definition and (b) the default option -minQuant that aggressively inlines.

Using -no-minQuant and rewriting the specs as follows enables me (in the current development build at least) to successfully validate your code. Note that a method used in the spec ought to be marked pure and that the way you wrote the ensures clause does not avoid an undefined case of rest.size() being called with rest == null. The following is better.

/* A simple linked list of entries. */
public class EntryList {
    /*@spec_public @*/ Object first;
    /*@spec_public @*/ EntryList rest;

    EntryList( Object first, EntryList rest) {
        this.first = first;
        this.rest = rest;
    }

    /*@ ensures (this.first == null ==> \result == 0);
      @ ensures (this.first != null && this.rest == null) ==> \result == 1;
      @ ensures (this.first != null && this.rest != null) ==> \result == (rest.size() + 1);
      @*/
    //@ pure
    public int size() {
        if(this.first == null) {
            return 0;
        }
        if(this.rest == null) {
            return 1;
        } else {
            return 1 + rest.size();
        }
    }
}

**Comment by davidcok** [comment]
Tuning this a bit more. If nullable is not the default, then some nullable annotations are needed:

public class EntryList {
    /*@spec_public nullable @*/ Object first;
    /*@spec_public nullable @*/ EntryList rest;

    EntryList( /*@ nullable */ Object first,  /*@ nullable */ EntryList rest) {
        this.first = first;
        this.rest = rest;
    }

    /*@ ensures (this.first == null && \result == 0) ||(this.rest == null && \result == 1)  ||
      @ (\result == rest.size() + 1);
      @ pure
      @*/
    public int size() {
        if(this.first == null) {
            return 0;
        }
        if(this.rest == null) {
            return 1;
        } else {
            return 1 + rest.size();
        }
    }
}

**Comment by davidcok** [comment]
Fixed for V0.8.23
